src/awsdata.jl

```diff
@@ -0,0 +1,112 @@
+using AWSS3
+using AWSS3: AWSConfig
+
+const allen_config = AWSConfig(; creds=nothing, region="us-west-2")
+
+# Create a few types that allow JSON3 to parse the manifest.json file
+struct AWSDir
+    relative_path::String
+    version::String
+    total_size::Int
+    url::String
+    view_link::String
+end
+Base.show(io::IO, d::AWSDir) = print(io, "AWSDir(\"", d.relative_path, "\", size=", d.total_size, ")")
+
+struct AWSDirs
+    directories::Dict{String,AWSDir}
+end
+Base.show(io::IO, d::AWSDirs) = print(io, [k => d for (k, d) in d.directories])
+Base.getindex(d::AWSDirs, k) = d.directories[k]
+Base.iterate(d::AWSDirs) = iterate(d.directories)
+Base.iterate(d::AWSDirs, i) = iterate(d.directories, i)
+
+struct AWSManifest
+    resource_uri::String
+    version::String
+    directory_listing::Dict{String,AWSDirs}
+    file_listing::Dict{String,Any}
+end
+Base.show(io::IO, manifest::AWSManifest) = print(io, "AWSManifest(version = \"", manifest.version,
+    "\", length(directory_listing) = ", length(manifest.directory_listing),
+    ", length(file_listing) = ", length(manifest.file_listing), ")")
+
+StructTypes.StructType(::Type{AWSDir}) = StructTypes.Struct()
+StructTypes.StructType(::Type{AWSDirs}) = StructTypes.Struct()
+StructTypes.StructType(::Type{AWSManifest}) = StructTypes.Struct()
+
+function awsmanifest(version)
+    url = "https://allen-brain-cell-atlas.s3-us-west-2.amazonaws.com/releases/$version/manifest.json"
+    rq = HTTP.request("GET", url)
+    return JSON3.read(String(rq.body), AWSManifest)
+end
+
+bucket(manifest::AWSManifest) = startswith(manifest.resource_uri, "s3://") ? split(manifest.resource_uri, "/")[3] : error("Not an S3 URI: ", manifest.resource_uri)
+
+"""
+    download_dir(manifest, relative_path, to; [config])
+
+Download Allen Brain data form `manifest.resource_uri * relative_path` to `to`.
+
+# Positional arguments:
+
+- `manifest`: Allen Brain manifest data
+
+    You can get with call `version = "20231215"; manifest = awsmanifest(version)`.
+
+- `relative_path`: Relative path of the data you want to download. Full path will be constructed with `manifest.resource_uri * relative_path`.
+
+- `to` : Download path and file name
+
+# Keyword arguments:
+
+- `config`: configuratoin to connect Allen Brain dataset location in AWS.
+
+
+# Examples
+To download data with feature_matrix_label = "WMB-10Xv2-TH" which is in dataset_label="WMB-10Xv2"
+
+```jldoctest
+
+julia> expression_matrices = manifest.file_listing["WMB-10Xv2"]["expression_matrices"]
+
```

Comment:
I'm not sure if the blank line between the input (`julia> expr`) and the output is something your editor added, but to me it makes it harder to figure out what output goes with what input. Can we delete it, here and below? (A blank line between the output and the next input is, of course, normal and desirable.)

Reply:
changed